query-translator/src/translate.rs

```diff
@@ -157,10 +159,53 @@ impl ToConstraint for ColumnConstraint {
                     right: right.into(),
                 }
             },
+            HasTypes { value: table, value_types, strict: strict_requested } => {
+                // If strict mode checking is on, and need to check exactly 1
+                // (not both or neither) of ValueType::Double and ValueType::Long
+                // we emit a Constraint::TypeCheck.
+                let num_numeric_checks = (value_types.contains(ValueType::Double) as i32) +
+                                         (value_types.contains(ValueType::Long) as i32);
+                let strict = strict_requested && num_numeric_checks == 1;
```

Comment:
You're doing the numeric checks even if `!strict_requested`!

So:

```
let strict = strict_requested &&
             value_types.is_unit() &&
             …;
```

Reply:
This bothers me a little, because you're imposing here some knowledge about the precise SQL encoding that we use for types, and that logic is mostly elsewhere. If we add, say, a geospatial floating-point representation — different type tag but `Double` internally — is something going to go wrong here? Can we make this more abstract?

Is the abstract thing you're asking: do we need to check the type affinity of the value to determine whether the value matches our requested value type set?

I think you're saying that for {Double, Integer} we do not need to check the type affinity, because only Double and Integer share a type tag, and if they're both present then the value types correspond to the entire value space.

And so the inverse, which you're doing here, is: are all of the _types_ that are represented by the type tags that are derived from our `value_types` present in `value_types`?

We use `SQLValueTypeSet` to hide all the type-tag-related stuff, so imagine something like:

```
use mentat_core::SQLValueTypeSet;
…

// Ridiculously expensive, but doesn't include its own type tag manipulation.
fn requires_type_affinity_check(value_types: ValueTypeSet) -> bool {
    let mut type_set = ValueTypeSet::none();
    for tag in value_types.iter().map(|vt| vt.value_type_tag()) {
        // from_tag doesn't exist.
        type_set = type_set.union(SQLValueTypeSet::from_tag(tag));
    }
    !type_set.is_strict_subset(value_types)
}
```
    
I'm not suggesting you write exactly that code, but I'd like you to think about what's happening here, and move whatever relates to type tags into `SQLValueType` and friends. This function here should just check a predicate.

Reply:
> If we add, say, a geospatial floating-point representation — different type tag but Double internally — is something going to go wrong here?

No, I think that should work. If you added arbitrary precision integer type that had tag = 5 but was stored as a blob, it would be broken though. (Admittedly, you'll loose free comparisons, so I don't think this would be a good idea, but then again, I don't really think using the same tag for Double and Long is a good idea either).

> We use SQLValueTypeSet to hide all the type-tag-related stuff, so imagine something like:

That's still not enough to get to the point where you could add new types that are only discriminated by their SQLite type affinity without modifying translate.rs. Note that a single predicate can never be enough, since you need to know which affinity to emit the check for.

What you end up needing is something like what I've done in the 2nd of my new commits. I'm not sure it's what you had in mind, I'm not a big fan of it since it's more complex (at least conceptually) and less efficient to support cases that can't happen given the way the types are currently set up. I also don't love exporting SQLTypeAffinity from core, but it seems hard to avoid.